test/relayclient/RelayProvider.test.ts

```diff
@@ -0,0 +1,274 @@
+    const paymasterInstance = await TestPaymasterEverythingAccepted.new()
+    paymaster = paymasterInstance.address
+    await paymasterInstance.setHub(relayHub.address)
+    await paymasterInstance.deposit({ value: web3.utils.toWei('2', 'ether') })
+    relayProcess = await startRelay(relayHub.address, stakeManager, {
+      stake: 1e18,
+      url: 'asd',
+      relayOwner: accounts[1],
+      EthereumNodeUrl: underlyingProvider.host
+    })
+  })
+
+  after(async function () {
+    await stopRelay(relayProcess)
+  })
+
+  it('should relay transparently', async function () {
+    const TestRecipient = artifacts.require('TestRecipient')
+    const testRecipient = await TestRecipient.new()
+    const forwarder = await testRecipient.getTrustedForwarder()
+    const gsnConfig = configureGSN({ relayHubAddress: relayHub.address })
+    const relayProvider = new RelayProvider(undefined, underlyingProvider, gsnConfig)
+    // NOTE: in real application its enough to set the provider in web3.
+    // however, in Truffle, all contracts are built BEFORE the test have started, and COPIED the web3,
+    // so changing the global one is not enough.
+    // @ts-ignore
+    TestRecipient.web3.setProvider(relayProvider)
+    const res = await testRecipient.emitMessage('hello world', {
+      from: gasLess,
+      forceGasPrice: '0x51f4d5c00',
+      // TODO: for some reason estimated values are crazy high!
+      gas: '100000',
+      paymaster,
+      forwarder
+    })
+    assert.equal(res.logs[0].event, 'SampleRecipientEmitted')
+    assert.equal(res.logs[0].args.message, 'hello world')
+    assert.equal(res.logs[0].args.realSender, gasLess)
+  })
+
+  describe('_ethSendTransaction', function () {
+    const id = 777
+    let testRecipient: TestRecipientInstance
+    let gsnConfig: GSNConfig
+    let jsonRpcPayload: JsonRpcPayload
+    let dependencyTree: TmpDependencyTree
+
+    before(async function () {
+      const TestRecipient = artifacts.require('TestRecipient')
+      testRecipient = await TestRecipient.new()
+      const forwarder = await testRecipient.getTrustedForwarder()
+      gsnConfig = configureGSN({ relayHubAddress: relayHub.address })
+      dependencyTree = RelayClient.getDefaultDependencies(underlyingProvider, gsnConfig)
+      // call to emitMessage('hello world')
+      jsonRpcPayload = {
+        jsonrpc: '2.0',
+        id,
+        method: 'eth_sendTransaction',
+        params: [
+          {
+            from: gasLess,
+            gas: '0x186a0',
+            gasPrice: '0x4a817c800',
+            forceGasPrice: '0x51f4d5c00',
+            paymaster,
+            forwarder,
+            to: testRecipient.address,
+            data: '0x2ac0df260000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b68656c6c6f20776f726c64000000000000000000000000000000000000000000'
```

Comment:
why not "encodeABI" ?

Reply:
(that is: if you did you some calculation to get a value, then put the calculation, not the value. the same is for gasPrice: I'm sure you didn't decide on this string of hex digits. you probably had some integer value in mind, whcih is no longer visible

Reply:
gas price everywhere is 22M that is set in server as minimum, I think, but it can be just a random big enough number, it is not under test here